pandas/tests/indexing/multiindex/test_multiindex.py

```diff
@@ -271,3 +271,13 @@ def test_multiindex_assign_aligns_as_implicit_tuple(self):
         df1["C"] = s1
         tm.assert_frame_equal(df1, df2)
         tm.assert_frame_equal(df1, df3)
+
+    def test_multiindex_assign_alignment_with_non_string_dtype(self):
+        # GH 62518
+        meta = DataFrame(
+            columns=MultiIndex.from_arrays(
+                [["a", "a", "z", "z"], pd.Categorical([1, 2, 1, 2])]
+            ),
+            dtype=object,
+        )
+        meta["z"] = meta["z"].astype("int64")
```

Comment:
Sorry, one more thing. Could you use `tm.assert_frame.equal` to assert that `meta` is equal to a new DataFrame that's the result of this casting?